numba_dpex/core/utils/reduction_kernel_builder.py

```diff
@@ -0,0 +1,388 @@
+import copy
+import warnings
+
+from numba.core import types
+from numba.core.errors import NumbaParallelSafetyWarning
+from numba.core.ir_utils import (
+    add_offset_to_labels,
+    get_name_var_table,
+    get_unused_var_name,
+    legalize_names,
+    mk_unique_var,
+    remove_dels,
+    rename_labels,
+    replace_var_names,
+)
+from numba.core.typing import signature
+
+from numba_dpex import config
+
+from .kernel_builder import (
+    ParforKernel,
+    _compile_kernel_parfor,
+    _print_body,
+    _to_scalar_from_0d,
+)
+from .kernel_templates.partial_reduction_template import (
+    RemainderReduceIntermediateKernelTemplate,
+    TreeReduceIntermediateKernelTemplate,
+)
+from .reduction_helper import update_sentinel
+
+
+def create_reduction_main_kernel_for_parfor(
+    loop_ranges,
+    parfor_node,
+    typemap,
+    flags,
+    has_aliases,
+    reductionKernelVar,
+    parfor_reddict=None,
+):
+    """
+    Creates a numba_dpex.kernel function for a parfor node.
+
+    There are two parts to this function:
+
+        1) Code to iterate across the iteration space as defined by
+           the schedule.
+        2) The parfor body that does the work for a single point in
+           the iteration space.
+
+    Part 1 is created as Python text for simplicity with a sentinel
+    assignment to mark the point in the IR where the parfor body
+    should be added. This Python text is 'exec'ed into existence and its
+    IR retrieved with run_frontend. The IR is scanned for the sentinel
+    assignment where that basic block is split and the IR for the parfor
+    body inserted.
+    """
+
+    redarrs = reductionKernelVar.parfor_redvars_to_redarrs
```

Comment:
Maybe convert into properties of `reductionKernelVar` and use them as such without creating local references. The loop_body property can return a deep copy.